Subject: [PPC,SOUND] Fix audio gpio state detection                                                                           
From: Andreas Schwab <schwab@suse.de>                                                                                         

When booting with line out or headphone plugged, you won't hear anything.
The problem is that after reset all channels are muted, but the actual
value of the gpio port doesn't exactly match the active_val settings as
expected by check_audio_gpio.  For example, the line_mute port is set to
7, but check_audio_gpio would expect 0xd or 0xf, thus its return value
indicates that it is not active, even though it is.  AFAICS only looking
at the low bit is enough to determine whether the port is active.  

Signed-off-by: Andreas Schwab <schwab@suse.de>



 sound/ppc/tumbler.c |    9 ++++++---
 1 files changed, 6 insertions(+), 3 deletions(-)

Index: linux-2.6.16-rc5-olh/sound/ppc/tumbler.c
===================================================================
--- linux-2.6.16-rc5-olh.orig/sound/ppc/tumbler.c
+++ linux-2.6.16-rc5-olh/sound/ppc/tumbler.c
@@ -191,7 +191,8 @@ static int check_audio_gpio(struct pmac_
 
 	ret = do_gpio_read(gp);
 
-	return (ret & 0xd) == (gp->active_val & 0xd);
+	printk("%s addr %x ret %x av %x\n",__FUNCTION__,gp->addr,ret,gp->active_val);
+	return (ret & 0x1) == (gp->active_val & 0x1);
 }
 
 static int read_audio_gpio(struct pmac_gpio *gp)
@@ -199,7 +200,9 @@ static int read_audio_gpio(struct pmac_g
 	int ret;
 	if (! gp->addr)
 		return 0;
-	ret = ((do_gpio_read(gp) & 0x02) !=0);
+	ret = do_gpio_read(gp);
+	printk("%s addr %x ret %x av %x\n",__FUNCTION__,gp->addr,ret,gp->active_val);
+	ret = (ret & 0x02) !=0;
 	return ret == gp->active_state;
 }
 
@@ -958,7 +961,7 @@ static void device_change_handler(void *
 	headphone = tumbler_detect_headphone(chip);
 	lineout = tumbler_detect_lineout(chip);
 
-	DBG("headphone: %d, lineout: %d\n", headphone, lineout);
+	printk("headphone: %d, lineout: %d\n", headphone, lineout);
 
 	if (headphone || lineout) {
 		/* unmute headphone/lineout & mute speaker */
